ui/src/message_popup/scaffolding_card.jsx

```diff
@@ -1,61 +1,149 @@
-    sessionLength: React.PropTypes.number.isRequired,
-    questions: React.PropTypes.array.isRequired,
-    onSliderChange: React.PropTypes.func.isRequired,
-    shouldShowStudentCard: React.PropTypes.bool.isRequired,
-    onStudentCardsToggled: React.PropTypes.func.isRequired,
-    shouldShowSummary: React.PropTypes.bool.isRequired,
-    onSummaryToggled: React.PropTypes.func.isRequired,
-    helpType: React.PropTypes.string.isRequired,
-    onHelpToggled: React.PropTypes.func.isRequired,
-    itemsToShow: React.PropTypes.object.isRequired
+    scaffolding: React.PropTypes.shape({
+      helpType: React.PropTypes.string.isRequired,
+      shouldShowStudentCard: React.PropTypes.bool.isRequired,
+      shouldShowSummary: React.PropTypes.bool.isRequired
+    }).isRequired,
+    email: React.PropTypes.string.isRequired,
+    itemsToShow: React.PropTypes.object.isRequired,
+    saveScaffoldingAndSession: React.PropTypes.func.isRequired
   },
   
+  getInitialState(){
+    const isSolutionMode = _.has(this.props.itemsToShow, 'solution');
+    return ({
+      email: this.props.email,
+      questions: allQuestions,
+      sessionLength: 10,
+      helpType: this.props.scaffolding.helpType,
+      shouldShowStudentCard: this.props.scaffolding.shouldShowStudentCard,
+      shouldShowSummary: this.props.scaffolding.shouldShowSummary,
+      competencyGroupValue: ALL_COMPETENCY_GROUPS,
+      isSolutionMode
+    });
+  },
+
+  getQuestions(competencyGroup=""){
+    var {competencyGroupValue, isSolutionMode} = this.state;
+    if(competencyGroup !== ""){
+      competencyGroupValue = competencyGroup;
+    }
+    const questions = (isSolutionMode || competencyGroupValue === ALL_COMPETENCY_GROUPS)
+      ? _.shuffle(this.withStudents(allQuestions))
+      : this.questionsForCompetencies(competencyGroupValue);
+    return questions;
+  },
+  
+  withStudents(questions) {
+    return questions.map((question) => {
+      const student = _.find(allStudents, {id: question.studentId });
+      return _.extend({student}, question);
+    });
+  },
+  
+  questionsForCompetencies(competencyGroup) {
+    const withCompetencyGroups = _.compact(allQuestions.map((question) => {
+      const learningObjective = _.find(learningObjectives, { id: question.learningObjectiveId });
+      if (learningObjective.competencyGroup !== competencyGroup) return null;
+      return {
+        ...question,
+        competencyGroup: learningObjective.competencyGroup
+      };
+    }));
+    return _.shuffle(this.withStudents(withCompetencyGroups));
+  },
+
+  onSave(){
+    const scaffolding = {
+      helpType: this.state.helpType,
+      shouldShowStudentCard: this.state.shouldShowStudentCard,
+      shouldShowSummary: this.state.shouldShowSummary
+    };
+    const gameSession = {
+      email: this.state.email,
+      sessionLength: this.state.sessionLength,
+      questions: this.getQuestions(this.state.competencyGroupValue)
```

Comment:
I don't think any other components need to know about `sessionLength` - ultimately it's only used to produce a list of questions for the gameSession.  If we do something like `questions.slice(0, sessionLength)` then we can pass back just the `email` and the list of questions to the parent.